testing/integration/upgrade_rollback_test.go

```diff
@@ -0,0 +1,195 @@
+
+	// Configure Agent with fast watcher configuration and also an invalid
+	// input when the Agent version matches the upgraded Agent version. This way
+	// the pre-upgrade version of the Agent runs healthy, but the post-upgrade
+	// version doesn't.
+	preInstallHook := func() error {
+		invalidInputPolicy := upgradetest.FastWatcherCfg + fmt.Sprintf(`
+outputs:
+  default:
+    type: elasticsearch
+    hosts: [127.0.0.1:9200]
+
+inputs:
+  - condition: '${agent.version.version} == "%s"'
+    type: invalid
+    id: invalid-input
+`, upgradeToVersion)
+		return startFixture.Configure(ctx, []byte(invalidInputPolicy))
+	}
+
+	// Use the post-upgrade hook to bypass the remainder of the PerformUpgrade
+	// because we want to do our own checks for the rollback.
+	var ErrPostExit = errors.New("post exit")
+	postUpgradeHook := func() error {
+		return ErrPostExit
+	}
+
+	err = upgradetest.PerformUpgrade(
+		ctx, startFixture, endFixture, t,
+		upgradetest.WithPreInstallHook(preInstallHook),
+		upgradetest.WithPostUpgradeHook(postUpgradeHook))
+	if !errors.Is(err, ErrPostExit) {
+		require.NoError(t, err)
+	}
+
+	// rollback should now occur
+
+	// wait for the agent to be healthy and back at the start version
+	err = upgradetest.WaitHealthyAndVersion(ctx, startFixture, startVersionInfo.Binary, 10*time.Minute, 10*time.Second, t)
+	if err != nil {
+		// agent never got healthy, but we need to ensure the watcher is stopped before continuing
+		// this kills the watcher instantly and waits for it to be gone before continuing
+		watcherErr := upgradetest.WaitForNoWatcher(ctx, 1*time.Minute, time.Second, 100*time.Millisecond)
+		if watcherErr != nil {
+			t.Logf("failed to kill watcher due to agent not becoming healthy: %s", watcherErr)
+		}
+	}
+	require.NoError(t, err)
+
+	// rollback should stop the watcher
+	// killTimeout is greater than timeout as the watcher should have been
+	// stopped on its own, and we don't want this test to hide that fact
+	err = upgradetest.WaitForNoWatcher(ctx, 2*time.Minute, 10*time.Second, 3*time.Minute)
+	require.NoError(t, err)
+
+	// now that the watcher has stopped lets ensure that it's still the expected
+	// version, otherwise it's possible that it was rolled back to the original version
+	err = upgradetest.CheckHealthyAndVersion(ctx, startFixture, startVersionInfo.Binary)
+	assert.NoError(t, err)
+}
+
+// TestStandaloneUpgradeRollbackOnRestarts tests the scenario where upgrading to a new version
+// of Agent fails due to the new Agent binary not starting up. It checks that the Agent is
+// rolled back to the previous version.
+func TestStandaloneUpgradeRollbackOnRestarts(t *testing.T) {
+	define.Require(t, define.Requirements{
+		Local: false, // requires Agent installation
+		Sudo:  true,  // requires Agent installation
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Start at the build version as we want to test the retry
+	// logic that is in the build.
```

Comment:
I think this comment is not relevant to this test?

Reply:
It is relevant.